emulator/transaction-emulator.cpp

```diff
@@ -0,0 +1,221 @@
+                                                         block::StoragePhaseConfig* storage_phase_cfg,
+                                                         block::ComputePhaseConfig* compute_phase_cfg,
+                                                         block::ActionPhaseConfig* action_phase_cfg) {
+  bool external{false}, ihr_delivered{false}, need_credit_phase{false};
+
+  if (msg_root.not_null()) {
+    auto cs = vm::load_cell_slice(msg_root);
+    external = block::gen::t_CommonMsgInfo.get_tag(cs);
+  }
+
+  if (trans_type == block::transaction::Transaction::tr_ord) {
+    need_credit_phase = !external;
+  } else if (trans_type == block::transaction::Transaction::tr_merge_install) {
+    need_credit_phase = true;
+  }
+
+  std::unique_ptr<block::transaction::Transaction> trans =
+      std::make_unique<block::transaction::Transaction>(*acc, trans_type, lt, utime, msg_root);
+
+  if (msg_root.not_null() && !trans->unpack_input_msg(ihr_delivered, action_phase_cfg)) {
+    if (external) {
+      // inbound external message was not accepted
+      return td::Status::Error(-701,"inbound external message rejected by account "s + acc->addr.to_hex() +
+                                                           " before smart-contract execution");
+    }
+    return td::Status::Error(-669,"cannot unpack input message for a new transaction");
+  }
+
+  if (trans->bounce_enabled) {
+    if (!trans->prepare_storage_phase(*storage_phase_cfg, true)) {
+      return td::Status::Error(-669,"cannot create storage phase of a new transaction for smart contract "s + acc->addr.to_hex());
+    }
+    if (need_credit_phase && !trans->prepare_credit_phase()) {
+      return td::Status::Error(-669,"cannot create credit phase of a new transaction for smart contract "s + acc->addr.to_hex());
+    }
+  } else {
+    if (need_credit_phase && !trans->prepare_credit_phase()) {
+      return td::Status::Error(-669,"cannot create credit phase of a new transaction for smart contract "s + acc->addr.to_hex());
+    }
+    if (!trans->prepare_storage_phase(*storage_phase_cfg, true, need_credit_phase)) {
+      return td::Status::Error(-669,"cannot create storage phase of a new transaction for smart contract "s + acc->addr.to_hex());
+    }
+  }
+
+  if (!trans->prepare_compute_phase(*compute_phase_cfg)) {
+    return td::Status::Error(-669,"cannot create compute phase of a new transaction for smart contract "s + acc->addr.to_hex());
+  }
+
+  if (!trans->compute_phase->accepted) {
+    if (external) {
+      // inbound external message was not accepted
+      auto const& cp = *trans->compute_phase;
+      return td::Status::Error(
+          -701,
+          PSLICE() << "inbound external message rejected by transaction " << acc->addr.to_hex() << ":\n" <<
+              "exitcode=" << cp.exit_code << ", steps=" << cp.vm_steps << ", gas_used=" << cp.gas_used <<
+              (cp.vm_log.empty() ? "" : "\nVM Log (truncated):\n..." + cp.vm_log));
```

Comment:
`cp.vm_log` will be empty by default.
You can add `compute_phase_cfg.with_vm_log = true;` to enable it.